solidity/random-beacon/contracts/libraries/Groups.sol

```diff
@@ -148,29 +127,194 @@ library Groups {
         );
     }
 
-    // TODO: Add group termination and expiration
+    /// @notice Goes through groups starting from the oldest one that is still
+    ///         active and checks if it hasn't expired. If so, updates the information
+    ///         about expired groups so that all expired groups are marked as such.
+    function expireOldGroups(Data storage self) internal {
+        // Move expiredGroupOffset as long as there are some groups that should
+        // be marked as expired. It is possible that expired group offset will
+        // move out of the groups array by one position. It means that all groups
+        // are expired (it points to the first active group) and that place in
+        // groups array - currently empty - will be possibly filled later by
+        // a new group.
+        while (
+            self.expiredGroupOffset < self.groupsRegistry.length &&
+            groupLifetimeOf(
+                self,
+                self.groupsRegistry[self.expiredGroupOffset]
+            ) <
+            block.number
+        ) {
+            Group memory group = self.groupsData[
+                self.groupsRegistry[self.expiredGroupOffset]
+            ];
+            // Decrement only for non-terminated groups, because terminateGroup()
+            // function handles 'activeGroupsCount' counter for terminated groups.
+            if (!group.terminated) {
+                self.activeGroupsCount--;
+            }
```

Comment:
>  I do not see us doing this check in v1, why is it needed?

It is needed because just like you said in your second comment we used `self.groupRegistry.length - self.expiredGroupOffset - self.activeTerminatedGroups.length`. If a group was terminated and then it was expired, we do not want to decrement the `self.activeGroupsCount` counter 2 times for the same group no matter if it's `terminated` or `expired`. I tried to use the original approach that was used in v1, but it bite me in other tests and the reason is that it is coupled with `expiredGroupOffset`. In other words, the only place where we move `expiredGroupOffset` is `expireOldGroups()`, so every time we call `activeTerminatedGroups()` without `expireOldGroups()` it may return an outdated result, cause the blocks already moved forward.

Reply:
I am still not following. I think we shouldn't have `self.activeGroupsCount` stored as a field and incrementing/decrementing it. Instead, we should have a function and compute it on the fly as `self.groupRegistry.length - self.expiredGroupOffset - self.activeTerminatedGroups.length`.

In `expireOldGroups` we first move the offset and in the same function, in the same transaction, we update `activeTerminatedGroups` based on the updated offset. When `expireOldGroups` finished, we should not have expired group in `activeTerminatedGroups` so I think `self.groupRegistry.length - self.expiredGroupOffset - self.activeTerminatedGroups.length` holds.

> activeTerminatedGroups() without expireOldGroups() it may return an outdated result, cause the blocks already moved forward.

Correct, and that's why no one but `Groups` library should use `activeTerminatedGroups`. This is similar situation as with `activeGroupsCount()` - it returns the number of groups based on the current stored state, and the stored state may be outdated. The group is not actually expired until the offset is updated on the chain.

Can you give some concrete example when it does not work? This would help me understand what the problem is.

Reply:
> Can you give some concrete example when it does not work?

We have two tests that are failing if we change a counter with what we had in v1, but they are failing for the same reason. It boils down to at what point we consider a group to be "active". We have 2 functions: `addCandidateGroup` & `activateCandidateGroup` and in the current approach, we consider it to be active when `activateCandidateGroup()` is called. In `addCandidateGroup()` we populate `groupsData` map & `groupsRegistry` array, but increment the counter in `activateCandidateGroup()`.
Here are the 2 tests that will fail [here](https://github.com/keep-network/keep-core/blob/main/solidity/random-beacon/test/Groups.test.ts#L597) and [here](https://github.com/keep-network/keep-core/blob/main/solidity/random-beacon/test/Groups.test.ts#L695). Take a look at the `beforeEach` where we add and activate groups. In the v1 approach, we analyzed if the group was active or not just by looking at the `groupsRegistry` and we didn't have a notion of separately activating a group. And that's the difference.
`activateCandidateGroup()` is used by `approveDkgResult()`. Not entirely sure how it can impact the functionality down the road if we decide to change it.

Reply:
Actually, we have another test that will [here](https://github.com/keep-network/keep-core/blob/main/solidity/random-beacon/test/RandomBeacon.GroupCreation.test.ts#L172). It expects "current state is not IDLE" but was reverted with "not awaiting genesis"

Reply:
I think we could solve it by one additional condition in `numberOfActiveGroups()`. After we evaluate the number with `self.groupRegistry.length - self.expiredGroupOffset - self.activeTerminatedGroups.length`, we need to check if the last group in `self.groupRegistry` has non-zero activation field. If not, we decrease the result by one.

I think this is safer than maintaining the count separately and making sure it's properly updated in all cases. Also, we don't have to read the whole structure when expiring groups. With several groups to expire this could get expensive.

Reply:
I had to add a couple of extra checks to use a v1 approach, but overall the idea of checking if the last group is active makes sense. Also had to change tests, that added a couple of groups and activated only the last one, which from my knowledge is not possible. eb46440